File: sdk/src/lib/contracts/EIP1559GasProvider.ts
import { JsonRpcProvider } from "@ethersproject/providers";
import { BigNumber } from "ethers";
import { FeeEstimationError } from "../utils/errors";
import { DEFAULT_GAS_ESTIMATION_PERCENTILE, DEFAULT_MAX_FEE_PER_GAS } from "../utils/constants";

type Fees = {
  maxFeePerGas: BigNumber;
  maxPriorityFeePerGas?: BigNumber;
};

type FeeHistory = {
  oldestBlock: number;
  reward: string[][];
  baseFeePerGas: string[];
  gasUsedRatio: number[];
};

export class EIP1559GasProvider {
  private feesCacheOld: Fees;
  private feesCache: Fees;
  private cacheIsValidForBlockNumber: BigNumber;
  private maxFeePerGasFromConfig: BigNumber;
  private gasEstimationPercentile: number;

  constructor(protected readonly provider: JsonRpcProvider, maxFeePerGas?: number, gasEstimationPercentile?: number) {
    this.maxFeePerGasFromConfig = BigNumber.from(maxFeePerGas ?? DEFAULT_MAX_FEE_PER_GAS);
    this.gasEstimationPercentile = gasEstimationPercentile ?? DEFAULT_GAS_ESTIMATION_PERCENTILE;
    this.cacheIsValidForBlockNumber = BigNumber.from(0);
    this.feesCacheOld = { maxFeePerGas: this.maxFeePerGasFromConfig };
    this.feesCache = { maxFeePerGas: this.maxFeePerGasFromConfig };
  }

  public async get1559Fees(percentile = this.gasEstimationPercentile): Promise<Fees> {
    const currentBlockNumber = await this.provider.getBlockNumber();
    if (this.cacheIsValidForBlockNumber.lt(currentBlockNumber)) {
      const { reward, baseFeePerGas }: FeeHistory = await this.provider.send("eth_feeHistory", [
        "0x4",
        "latest",
        [percentile],
      ]);

      const maxPriorityFeePerGas = reward
        .reduce((acc: BigNumber, currentValue: string[]) => acc.add(currentValue[0]), BigNumber.from(0))
        .div(reward.length);

      if (maxPriorityFeePerGas && maxPriorityFeePerGas.gt(this.maxFeePerGasFromConfig)) {
        throw new FeeEstimationError(
          `Estimated miner tip of ${maxPriorityFeePerGas} exceeds configured max fee per gas of ${this.maxFeePerGasFromConfig}!`,
        );
      }

      this.cacheIsValidForBlockNumber = BigNumber.from(currentBlockNumber);

      const maxFeePerGas = BigNumber.from(baseFeePerGas[baseFeePerGas.length - 1])
        .mul(2)
        .add(maxPriorityFeePerGas);

      if (maxFeePerGas.gt(0) && maxPriorityFeePerGas.gt(0)) {
        this.feesCache = {
          maxPriorityFeePerGas,
          maxFeePerGas: maxFeePerGas.gt(this.maxFeePerGasFromConfig) ? this.maxFeePerGasFromConfig : maxFeePerGas,
        };
      } else {
        this.feesCache = {
          maxFeePerGas: this.maxFeePerGasFromConfig,
        };
      }
    }
    return this.feesCache;
  }

  public async get1559FeesOld(percentile = this.gasEstimationPercentile): Promise<Fees> {
    const currentBlockNumber = await this.provider.getBlockNumber();
    if (this.cacheIsValidForBlockNumber.lt(currentBlockNumber)) {
      const feeHistory: FeeHistory = await this.provider.send("eth_feeHistory", ["0xa", "pending", [percentile]]);

      const maxPriorityFeePerGas = this.calculatePriorityFee(feeHistory);

      const maxFeePerGas = BigNumber.from(feeHistory.baseFeePerGas[feeHistory.baseFeePerGas.length - 1])
        .mul(2)
        .add(maxPriorityFeePerGas);

      if (maxPriorityFeePerGas.gt(this.maxFeePerGasFromConfig)) {
        throw new FeeEstimationError(
          `Estimated miner tip of ${maxPriorityFeePerGas} exceeds configured max fee per gas of ${this.maxFeePerGasFromConfig}!`,
        );
      }

      this.cacheIsValidForBlockNumber = BigNumber.from(currentBlockNumber);

      if (maxFeePerGas.gt(0) && maxPriorityFeePerGas.gt(0)) {
        this.feesCacheOld = {
          maxFeePerGas: maxFeePerGas.gt(this.maxFeePerGasFromConfig) ? this.maxFeePerGasFromConfig : maxFeePerGas,
          maxPriorityFeePerGas: maxPriorityFeePerGas,
        };
      } else {
        this.feesCacheOld = {
          maxFeePerGas: this.maxFeePerGasFromConfig,
        };
      }
    }
    return this.feesCacheOld;
  }

  private calculatePriorityFee(feeHistory: FeeHistory): BigNumber {
    const priorityFeeWeightedSum = feeHistory.reward.reduce(
      (a, v, index) => a + Number(v[0]) * feeHistory.gasUsedRatio[index],
      0,
    );
    const sumOfRatios = feeHistory.gasUsedRatio.reduce((a, v) => a + v, 0);

    if (sumOfRatios === 0) {
      return BigNumber.from(0);
    }

    return BigNumber.from(Math.round(priorityFeeWeightedSum / (sumOfRatios * feeHistory.gasUsedRatio.length)));
  }
}

File: coordinator/clients/smart-contract-client/src/main/kotlin/net/consensys/linea/contract/EIP1559GasProvider.kt
package net.consensys.linea.contract

import net.consensys.linea.web3j.AtomicContractEIP1559GasProvider
import net.consensys.linea.web3j.EIP1559GasFees
import net.consensys.linea.web3j.blocksRange
import net.consensys.toIntervalString
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.web3j.protocol.Web3j
import org.web3j.protocol.core.DefaultBlockParameterName
import java.math.BigDecimal
import java.math.BigInteger

class EIP1559GasProvider(private val l2Client: Web3j, private val config: Config) :
  AtomicContractEIP1559GasProvider {
  private val log: Logger = LogManager.getLogger(this::class.java)

  data class Config(
    val gasLimit: BigInteger,
    val maxFeePerGasCap: BigInteger,
    val feeHistoryBlockCount: UInt,
    val feeHistoryRewardPercentile: Double
  )

  private val chainId: Long = l2Client.ethChainId().send().chainId.toLong()
  private var cacheIsValidForBlockNumber: BigInteger = BigInteger.ZERO
  private var feesCache: EIP1559GasFees = getRecentFees()
  private var gasLimitOverride = BigInteger.valueOf(Long.MAX_VALUE)

  private fun getRecentFees(): EIP1559GasFees {
    val currentBlockNumber = l2Client.ethBlockNumber().send().blockNumber
    if (currentBlockNumber > cacheIsValidForBlockNumber) {
      l2Client
        .ethFeeHistory(
          config.feeHistoryBlockCount.toInt(),
          DefaultBlockParameterName.LATEST,
          listOf(config.feeHistoryRewardPercentile)
        )
        .sendAsync()
        .thenApply {
            feeHistoryResponse ->
          var maxPriorityFeePerGas = feeHistoryResponse.feeHistory.reward.map { BigDecimal(it[0]) }
            .reduce { acc, reward ->
              acc.add(reward)
            }.divide(BigDecimal(feeHistoryResponse.feeHistory.reward.size)).toBigInteger()

          if (maxPriorityFeePerGas > config.maxFeePerGasCap) {
            maxPriorityFeePerGas = config.maxFeePerGasCap
            log.warn(
              "Estimated miner tip of $maxPriorityFeePerGas exceeds configured max " +
                "fee per gas of ${config.maxFeePerGasCap} returning cap instead!"
            )
          }

          cacheIsValidForBlockNumber = currentBlockNumber

          val maxFeePerGas = feeHistoryResponse.feeHistory.baseFeePerGas.last()
            .multiply(BigInteger.TWO)
            .add(maxPriorityFeePerGas)

          if (maxFeePerGas > BigInteger.ZERO && maxPriorityFeePerGas > BigInteger.ZERO) {
            feesCache = EIP1559GasFees(
              maxPriorityFeePerGas = maxPriorityFeePerGas,
              maxFeePerGas = maxFeePerGas.min(config.maxFeePerGasCap)
            )
            log.debug(
              "New fees estimation: fees={} l2Blocks={}",
              feeHistoryResponse.feeHistory.blocksRange().toIntervalString(),
              feesCache
            )
          } else {
            feesCache = EIP1559GasFees(
              maxPriorityFeePerGas = BigInteger.ZERO,
              maxFeePerGas = config.maxFeePerGasCap
            )
          }
        }
        .get()
    }
    return feesCache
  }

  override fun getGasPrice(contractFunc: String?): BigInteger {
    return getGasPrice()
  }

  @Deprecated("Deprecated in Java")
  override fun getGasPrice(): BigInteger {
    throw NotImplementedError("EIP1559GasProvider only implements EIP1559 specific methods")
  }

  fun overrideNextGasLimit(gasLimit: BigInteger) {
    gasLimitOverride = gasLimit
  }

  private fun resetGasLimitOverride() {
    gasLimitOverride = BigInteger.valueOf(Long.MAX_VALUE)
  }

  override fun getGasLimit(contractFunc: String?): BigInteger {
    val gasLimit = gasLimitOverride.min(getGasLimit())
    resetGasLimitOverride()

    return gasLimit
  }

  @Deprecated("Deprecated in Java")
  override fun getGasLimit(): BigInteger {
    return config.gasLimit
  }

  override fun isEIP1559Enabled(): Boolean {
    return true
  }

  override fun getChainId(): Long {
    return chainId
  }

  override fun getMaxFeePerGas(contractFunc: String?): BigInteger {
    return getRecentFees().maxFeePerGas
  }

  override fun getMaxPriorityFeePerGas(contractFunc: String?): BigInteger {
    return getRecentFees().maxPriorityFeePerGas
  }

  override fun getEIP1559GasFees(): EIP1559GasFees {
    return getRecentFees()
  }
}
